core/storage/user/gae_models.py

```diff
@@ -97,6 +97,61 @@ def get_by_normalized_username(cls, normalized_username):
             cls.normalized_username == normalized_username).get()
 
 
+class ActivitiesCompletedByLearnerModel(base_models.BaseModel):
+    """Keeps track of all the explorations and collections completed by the
+    learner.
+
+    Instances of this class are keyed by the user id.
+    """
+    # IDs of all the explorations completed by the user.
+    completed_exp_ids = ndb.StringProperty(repeated=True, indexed=True)
+    # IDs of all the collections completed by the user.
+    completed_collection_ids = ndb.StringProperty(repeated=True, indexed=True)
+
+
+class IncompleteCollectionsModel(base_models.BaseModel):
+    """Keeps track of all the collections currently being completed by the
+    learner.
+
+    Instances of this class are keyed by the user id.
+    """
+    incomplete_collection_ids = ndb.StringProperty(
+        repeated=True, indexed=True)
+
+
+class IncompleteExplorationsModel(base_models.BaseModel):
+    """Keeps track of all the explorations currently being completed by the
+    learner.
+
+    Instances of this class are keyed by the user id.
+    """
+    # A list which stores the explorations partially completed by the learner.
+    # Each item in the list is a json object keyed by exploration id and the
+    # value corresponds to a few extra details -
+    # timestamp: When did the learner leave the exploration.
+    # version: Version of the exploration.
+    # state_name: The name of the state at which the learner left the
+    # exploration. Example -
+    # [
+    #  {
+    #   (exp_id_1): {
+    #    'timestamp': timestamp,
```

Comment:
I have followed the second suggestion here. I have maintained this just to be a list of exploration ids and I am maintaining the "last playthrough" information separately. This way I can use incomplete exploration model just like all the other models and it is very easy to add new fields.